### ISSUE TITLE
[research] Support Metamask onboarding

### ISSUE THREAD

**Issue by Atmosfearful** [KlimaDAO/klimadao#567]
https://docs.metamask.io/guide/onboarding-library.html#examples

Does this work with web3modal?

If yes, we should implement

**Comment by biwano** [comment]
Hi @Atmosfearful!

I looked a little bit. I implemented it as explained in the tutorial, with a few changes so it works with server side rendering (The MetaMaskOnboarding uses the window variable). The Draft PR is here: https://github.com/KlimaDAO/klimadao/pull/575

Here is what I found:
 - If metamask is not installed:
   - The onboarding button redirects properly the user to the extension installation page and installation goes smoothly
   - But! When completing installation the user is not redirected properly to the website, he has to close the metamask windows (extension installation page and metamask plugin windows) manually. (1)
   -  Then he can can click on connect to open the classic web3modal and connect the site to metamask (still manually)
 - If metamask is installed but not connected to the site:
   - MetaMaskOnboarding opens the metamask window asking for a connecion (in the code it is the parts with the accounts)
   - But! Once the connection is made this is not reflected to the app because the user still has to click on connect to connect the site to metamask through web3 modal (2)

For point (1): I suspect there maybe aconflict between web3modal or ssr and MetaMaskOnboarding I would like to try with an empty nextjs project (or maybe a bug in the onboarding library)
For point (2) : I am not too familiar with web3modal but maybe there is a possibility to initialize web3modal with the MetaMaskOnboarding provider.

**Comment by Atmosfearful** [comment]
Thanks for looking into this @biwano 
Indeed I would very much prefer if this was integrated right into the modal.
Could you do more research into that? https://github.com/WalletConnect/web3modal

Keep in mind: the modal normally doesn't show metamask as an option, if you don't have metamask installed.

I think the logic would be straightforward:
if no metamask is present, inject a "create new metamask" onboarding button

**Comment by biwano** [comment]
Showing the onboarding button right into web3modal should be possible by creating a dummy provider. The workflow would be the following:
- The user clicks the button
- The button registers that the onbording process has started (html5 storage?) and triggers the onboarding library
- The user installs metamask
- The onboarding library reloads the Klima page. (meanwhile the user is configuring Metamask)
- The app detects the onboarding process has started, and triggers the web3modal with the legit Metamask provider. We cannot directly connect the wallet because the account is not configured yet. (meanwhile the user is still configuring metamask)
- The user finishes configuring Metamask, closes the Metamask and Metamask installation windows.
- The user is presented with the Klima app with the connection modal open.

We are supposed to be able to detect account changes and connect the app to Metamask with the connectTo option upon first account creation by the user. But it did not work in my tests.
There is an edge case if the user installs metamask but does not create an account. In this case, If the user tries to connect the app with metamask, he gets a "user rejected" error while the Metamask account creation wizard start. But I think it is already the case.

**Comment by biwano** [comment]
Hi @Atmosfearful 

I verified that what I said above was feasible in PR #575 

There are 3 problems I could not solve:
- Metamask does not redirects to the Klimadao site at the end of the process (I tested with a fresh empty vanilla React app and the behaviour is the same)
- If the user installs Metamask but for some reason does not configure an account, Kliamadao app will return a User rejected error when the user tries to connect to Metamask (It is already the case today)
- It is not possible to detect account creation because it requires the user to connect Metamask to the site beforehand which is counter-intuitive.

I think:
- It still improves the user experience.
- Maybe it can be improved with some UX work. eg. Better text for the dummy provider, A help text or link when the site is refreshed after Metamask is installed to explain how a browser wallet works?
